walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
+       */
+      if (!Collections.disjoint(optionalAttendees, event.getAttendees()))
+        optionalAttendeeEvents.add(event);
+      else findConflictAndResolve(freeTime, new LinkedList<>(request.getAttendees()), event);
+    }
+
+  }
+
+  // Try checking for any time slots we can get for the meeting with optional attendees
+  private LinkedList<TimeRange> freeSlotsWithOptionalAttendees(
+      LinkedList<Event> optionalAttendeeEvents,
+      LinkedList<TimeRange> freeTime,
+      LinkedList<String> optionalAttendees) {
+    // Create a new OptionalFreeTime list so that we still have our time slots with mandatory
+    // attendees
+    LinkedList<TimeRange> optionalFreeTime = new LinkedList<>(freeTime);
+    /* Loop through all the events our optional attendees are attending and
+     * find free time slots we can have for our meeting 
+     */
+    for (Event event : optionalAttendeeEvents) {
+      findConflictAndResolve(optionalFreeTime, optionalAttendees, event);
+    }
+    /* If we can't find any free time slots with optional attendees
+     * then return the original freeTime slots
+     */
+    if (optionalFreeTime.isEmpty()) optionalFreeTime = freeTime;
+
+    return optionalFreeTime;
+  }
+
+  // This method finds the overlapping free time slot and event; altering the freeTime so that it
+  // doesn't overlap anymore.
+  private void findConflictAndResolve(
+      LinkedList<TimeRange> time, LinkedList<String> attendees, Event event) {
+    if (!Collections.disjoint(attendees, event.getAttendees())) {
+      // Go through all the free time slots and check from which one is the event overlapping
+      ListIterator<TimeRange> iterator = time.listIterator();
+      while (iterator.hasNext()) {
+        TimeRange freeSlot = iterator.next();
+        // When th eoverlapping free slot is found send our free time and event slot to get altered.
+        if (freeSlot.overlaps(event.getWhen())) {
+          updateFreeTime(freeSlot, event.getWhen(), time, iterator);
+          // After updating free time is done, check if the event ends in this free slots
+          if (event.getWhen().end() <= freeSlot.end()) break;
+        }
+      }
+    }
+  }
+
+  /**
+   * This method free time, free time slot and current event with our free time slot and the iterator
+   * and checks all conditions in which the event and time slot could overlap. 
+   * Then calculates the free time minus overlapping part. 
+   * It checks the new slot with the meeting slot and then adds or removes it. 
+   * (By removing/ setting or adding in the iterator).
+   */
+  private void updateFreeTime(
+      TimeRange freeSlot,
+      TimeRange eventSlot,
+      LinkedList<TimeRange> freeTime,
+      ListIterator<TimeRange> iterator) {
+
+    int freeSlotStart = freeSlot.start();
+    int freeSlotEnd = freeSlot.end();
+
+    int eventSlotStart = eventSlot.start();
+    int eventSlotEnd = eventSlot.end();
+
+    TimeRange newSlot;
```

Comment:
New slot is only used locally, so there's no need to declare it on the function level scope.

Reply:
Okay will change its scope to block level.